File: app/src/main/java/com/example/android/pingpongscorekeeper/components/PingPongGame.java
package com.example.android.pingpongscorekeeper.components;

import java.util.ArrayList;

public class PingPongGame
{
    public PingPongPlayer playerOne;
    public PingPongPlayer playerTwo;
    public PingPongPlayer previousSetFirstServingPlayer;
    public PingPongPlayer currentServingPlayer;
    private PingPongPlayer initialServingPlayer;

    public String playerOneName;
    public String playerTwoName;

    public ArrayList<PingPongSet> pingPongSets;

    private final static int minNumberOfPointsToWinASet = 11;
    private final static int numberOfPointsPlayerNeedsToBeatOpponent = 2;

    private int winningNumberOfSets;

    public PingPongGame(
            String playerOneName,
            String playerTwoName,
            int numSets,
            String servingPlayer
    )
    {
        this.playerOneName = playerOneName;
        this.playerTwoName = playerTwoName;
        pingPongSets = new ArrayList<>();
        winningNumberOfSets = numSets % 2 == 1 ? numSets/2 + 1 : numSets / 2;


        playerOne = new PingPongPlayer(playerOneName);
        playerTwo = new PingPongPlayer(playerTwoName);

        currentServingPlayer = servingPlayer
                .equals(PlayerEnum.PLAYER_ONE.toString()) ? playerOne : playerTwo;
        initialServingPlayer = currentServingPlayer;
        previousSetFirstServingPlayer = currentServingPlayer;
    }


    public void setPlayerOneName(String name)
    {
        playerOne.setName(name);
    }

    public void setPlayerTwoName(String name)
    {
        playerTwo.setName(name);
    }

    public void resetGame()
    {
        playerOne.reset();
        playerTwo.reset();

        pingPongSets = new ArrayList<>();

        currentServingPlayer = initialServingPlayer;
        previousSetFirstServingPlayer = currentServingPlayer;
    }

    public int getCurrentSetNumber()
    {
        return playerOne.getNumberOfSetsWon() + playerTwo.getNumberOfSetsWon();
    }

    public int getPlayerOneCurrentSetScore()
    {
        return playerOne.getCurrentSetScore();
    }

    public int getPlayerTwoCurrentSetScore()
    {
        return playerTwo.getCurrentSetScore();
    }

    public int getNumSetsPlayerOneWon()
    {
        return playerOne.getNumberOfSetsWon();
    }

    public int getNumSetsPlayerTwoWon()
    {
        return playerTwo.getNumberOfSetsWon();
    }

    public boolean hasPlayerOneWonMatch()
    {
        return hasPlayerWonMatch(playerOne);
    }

    public boolean hasPlayerTwoWonMatch()
    {
        return hasPlayerWonMatch(playerTwo);
    }

    public boolean hasPlayerWonMatch(PingPongPlayer player)
    {
        return player.getNumberOfSetsWon() == winningNumberOfSets;
    }

    private boolean hasPlayerWonCurrentSet(PingPongPlayer player, PingPongPlayer opponent)
    {
        return player.getCurrentSetScore() >= minNumberOfPointsToWinASet &&
                player.getCurrentSetScore() - opponent.getCurrentSetScore() >=
                        numberOfPointsPlayerNeedsToBeatOpponent;
    }

    public boolean isDeuce()
    {
        return playerOne.getCurrentSetScore() >= minNumberOfPointsToWinASet - 1 &&
                playerOne.getCurrentSetScore() == playerTwo.getCurrentSetScore();
    }

    public boolean isGameOver()
    {
        return hasPlayerOneWonMatch() || hasPlayerTwoWonMatch();
    }

    public void playerHasScored(PingPongPlayer player)
    {
        updateServingPlayer();
        player.incrementCurrentSetScore();
    }

    private void switchCurrentServingPlayer()
    {
        currentServingPlayer = currentServingPlayer == playerOne ? playerTwo : playerOne;
    }

    public void updateServingPlayer()
    {
        if(playerOne.getCurrentSetScore() < minNumberOfPointsToWinASet - 1 ||
                playerTwo.getCurrentSetScore() < minNumberOfPointsToWinASet - 1)
        {
            if(((playerOne.getCurrentSetScore() + playerTwo.getCurrentSetScore())  % 2 ) == 1)
            {
                switchCurrentServingPlayer();
            }
        }
        else
        {
            switchCurrentServingPlayer();
        }
    }

    public boolean hasPlayerWonCurrentSet(PingPongPlayer player)
    {
        PingPongPlayer opponentPlayer = player == playerOne  ? playerTwo : playerOne;
        return hasPlayerWonCurrentSet(player, opponentPlayer);
    }

    public void incrementNumberOfSetsWon(PingPongPlayer player)
    {
        pingPongSets.add(new PingPongSet(
                playerOne.getNumberOfSetsWon() + playerTwo.getNumberOfSetsWon() + 1,
                playerOne.getCurrentSetScore(), playerTwo.getCurrentSetScore()));

        player.incrementNumberOfSetsWon();


        switchStartingServerForSet();
    }

    private void switchStartingServerForSet() {
        if(previousSetFirstServingPlayer == playerOne)
        {
            currentServingPlayer = playerTwo;
            previousSetFirstServingPlayer = playerTwo;

        }
        else
        {
            currentServingPlayer = playerOne;
            previousSetFirstServingPlayer = playerOne;
        }
    }

    public void resetPlayersCurrentSetScore()
    {
        playerOne.resetCurrentSetScore();
        playerTwo.resetCurrentSetScore();
    }
}

File: app/src/main/java/me/dhivo/android/pingpongmatchtracker/components/PingPongGame.kt
package me.dhivo.android.pingpongmatchtracker.components

import java.util.ArrayList

class PingPongGame(
        var playerOneName: String,
        var playerTwoName: String,
        numSets: Int,
        private val minNumberOfPointsToWinASet: Int,
        servingPlayer: String,
        val playerOneId: Long,
        val playerTwoId: Long
) {
    var playerOne: PingPongPlayer
    var playerTwo: PingPongPlayer
    var previousSetFirstServingPlayer: PingPongPlayer
    var currentServingPlayer: PingPongPlayer
    private val initialServingPlayer: PingPongPlayer

    var pingPongSets: ArrayList<PingPongSet>

    private val winningNumberOfSets: Int

    val currentSetNumber: Int
        get() = playerOne.numberOfSetsWon + playerTwo.numberOfSetsWon

    val playerOneCurrentSetScore: Int
        get() = playerOne.currentSetScore

    val playerTwoCurrentSetScore: Int
        get() = playerTwo.currentSetScore

    val numSetsPlayerOneWon: Int
        get() = playerOne.numberOfSetsWon

    val numSetsPlayerTwoWon: Int
        get() = playerTwo.numberOfSetsWon

    val isDeuce: Boolean
        get() = playerOne.currentSetScore >= minNumberOfPointsToWinASet - 1
                && playerOne.currentSetScore == playerTwo.currentSetScore

    val isGameOver: Boolean
        get() = hasPlayerOneWonMatch() || hasPlayerTwoWonMatch()

    init {
        pingPongSets = ArrayList()
        winningNumberOfSets = if (numSets % 2 == 1) numSets / 2 + 1 else numSets / 2


        playerOne = PingPongPlayer(playerOneName)
        playerTwo = PingPongPlayer(playerTwoName)

        currentServingPlayer = if (servingPlayer == PlayerEnum.PLAYER_ONE.toString())
            playerOne
        else
            playerTwo
        initialServingPlayer = currentServingPlayer
        previousSetFirstServingPlayer = currentServingPlayer
    }

    fun resetGame() {
        playerOne.reset()
        playerTwo.reset()

        pingPongSets = ArrayList()

        currentServingPlayer = initialServingPlayer
        previousSetFirstServingPlayer = currentServingPlayer
    }

    private fun hasPlayerOneWonMatch(): Boolean {
        return hasPlayerWonMatch(playerOne)
    }

    private fun hasPlayerTwoWonMatch(): Boolean {
        return hasPlayerWonMatch(playerTwo)
    }

    fun hasPlayerWonMatch(player: PingPongPlayer): Boolean {
        return player.numberOfSetsWon == winningNumberOfSets
    }

    private fun hasPlayerWonCurrentSet(player: PingPongPlayer, opponent: PingPongPlayer): Boolean {
        return player.currentSetScore >= minNumberOfPointsToWinASet
                && player.currentSetScore - opponent.currentSetScore >= numberOfPointsPlayerNeedsToBeatOpponent
    }

    fun playerHasScored(player: PingPongPlayer) {
        updateServingPlayer()
        player.incrementCurrentSetScore()
    }

    private fun switchCurrentServingPlayer() {
        currentServingPlayer = if (currentServingPlayer == playerOne) playerTwo else playerOne
    }

    private fun updateServingPlayer() {
        if (playerOne.currentSetScore < minNumberOfPointsToWinASet - 1
                || playerTwo.currentSetScore < minNumberOfPointsToWinASet - 1) {
            if ((playerOne.currentSetScore + playerTwo.currentSetScore) % 2 == 1) {
                switchCurrentServingPlayer()
            }
        } else {
            switchCurrentServingPlayer()
        }
    }

    fun hasPlayerWonCurrentSet(player: PingPongPlayer): Boolean {
        val opponentPlayer = if (player == playerOne) playerTwo else playerOne
        return hasPlayerWonCurrentSet(player, opponentPlayer)
    }

    fun incrementNumberOfSetsWon(player: PingPongPlayer) {
        pingPongSets.add(PingPongSet(
                playerOne.numberOfSetsWon + playerTwo.numberOfSetsWon + 1,
                playerOne.currentSetScore, playerTwo.currentSetScore))

        player.incrementNumberOfSetsWon()


        switchStartingServerForSet()
    }

    private fun switchStartingServerForSet() {
        if (previousSetFirstServingPlayer == playerOne) {
            currentServingPlayer = playerTwo
            previousSetFirstServingPlayer = playerTwo

        } else {
            currentServingPlayer = playerOne
            previousSetFirstServingPlayer = playerOne
        }
    }

    fun resetPlayersCurrentSetScore() {
        playerOne.resetCurrentSetScore()
        playerTwo.resetCurrentSetScore()
    }

    companion object {
        private const val numberOfPointsPlayerNeedsToBeatOpponent = 2
    }
}
